### ISSUE TITLE
Deduplication: add Python script to generate SQL-runner playbook for one-time deduplication of Redshift schema

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/snowplow#3074]
It'd be great to develop a little Python script in the vein of [redshift-unload-copy.py](https://github.com/awslabs/amazon-redshift-utils/blob/master/src/UnloadCopyUtility/redshift-unload-copy.py) that does a one off deduplication process for a list of tables / all the tables in a particular schema.

We will recommend that users run this 2-3 weeks after deploying cross-batch dedupe, this will:

1. Clean out all duplicates from before the upgrade to cross-batch dedupe was put live
2. Clean out new dupes that might be generated post upgrade because the DynamoDB table starts off empty

**Comment by alexanderdean** [comment]
This also means we can de-schedule #3045

**Comment by chuwy** [comment]
This is a single missing ticket in R88 CHANGELOG.

**Comment by bogaert** [comment]
This was pushed back to a later release.

**Comment by alexanderdean** [comment]
Cleared milestone, please ensure this isn't in the CHANGELOG.

**Comment by chuwy** [comment]
@alexanderdean this isn't in CHANGELOG for sure. Do we want to make it R89?

**Comment by alexanderdean** [comment]
Moved to R89

**Comment by BenFradet** [comment]
is this still part of r89?

**Comment by alexanderdean** [comment]
No, pushing back!

**Comment by iaingray** [comment]
Is anyone working on this? I'd be interested to if not.

**Comment by BenFradet** [comment]
@iaingray nobody is working on this particular issue inside Snowplow at the moment so feel free to have a crack at it and open a PR :+1:

**Comment by iaingray** [comment]
Okay, just so I'm clear, the script should generate a playbook to run the following queries on whatever event/unstruct/context tables exist within the db it's run against?

https://github.com/snowplow/snowplow/tree/master/5-data-modeling/deduplication-queries

**Comment by alexanderdean** [comment]
Can you confirm @bogaert ?

**Comment by bogaert** [comment]
Hi @iaingray,

We started work on this several months ago but it was quickly put on hold. I believe we wrote most of the Python script that generates the deduplication queries and SQL Runner playbook (please confirm, @dilyand).

However, there is a critical piece of extra work that needs to happen before we can release this and that's to significantly improve the deduplication queries. The two main shortcomings right now are:

- poor deduplication of the child tables (naive algorithm); and
- no solution for synthetic duplicates.

The child tables are currently deduplicated by simply grouping the fields in each child table. This method cannot distinguish between natural and synthetic duplicates (for which we need the event fingerprint).

The way the queries deal with synthetic duplicates (same event ID, different event fingerprint) is by simply quarantining them. We need a way to re-assign event ID in *both* the events table and child tables simultaneously.

This makes it a significantly larger exercise and–given our improvements to the Hadoop-based deduplication process in Snowplow 86 and 88–we decided to put this on hold.

If you're interested in having a crack at it, feel free to do so! Happy to advise along the way.

**Comment by alexanderdean** [comment]
De-scheduling

**Comment by alexanderdean** [comment]
Closing
